pgxn/neon/walsender_hooks.c

```diff
@@ -0,0 +1,113 @@
+#include "walsender_hooks.h"
+#include "postgres.h"
+#include "fmgr.h"
+#include "access/xlogdefs.h"
+#include "replication/walsender.h"
+#include "access/xlog.h"
+#include "access/xlog_internal.h"
+#include "access/xlogreader.h"
+#include "miscadmin.h"
+#include "utils/wait_event.h"
+
+#include "neon_walreader.h"
+#include "walproposer.h"
+
+static NeonWALReader *wal_reader = NULL;
+extern XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
+extern bool GetDonorShmem(XLogRecPtr *donor_lsn);
+
+static XLogRecPtr NeonWALReadWaitForWAL(XLogRecPtr loc)
+{
+    while(!NeonWALReaderUpdateDonor(wal_reader))
+        pg_usleep(1000);
+
+    return WalSndWaitForWal(loc);
+}
+
+static int NeonWALPageRead(
```

Comment:
FYI I don't fully understand this code, but it should be ok since it's used only in logical replication

Reply:
Added some comments just in case too, but yes I'm not too worried about it